nav2_util/include/nav2_util/service_client.hpp

```diff
@@ -34,24 +40,48 @@ class ServiceClient
   * @brief A constructor
   * @param service_name name of the service to call
   * @param provided_node Node to create the service client off of
+  * @param spin_thread Whether to spin with a dedicated thread internally
   */
   explicit ServiceClient(
     const std::string & service_name,
-    const NodeT & provided_node)
+    const NodeT & provided_node, bool spin_thread = false)
   : service_name_(service_name), node_(provided_node)
   {
-    callback_group_ = node_->create_callback_group(
-      rclcpp::CallbackGroupType::MutuallyExclusive,
-      false);
-    callback_group_executor_.add_callback_group(callback_group_, node_->get_node_base_interface());
+    if(spin_thread) {
+      callback_group_ = node_->create_callback_group(
+        rclcpp::CallbackGroupType::MutuallyExclusive,
+        false);
+    }
+
     client_ = node_->template create_client<ServiceT>(
       service_name,
       rclcpp::SystemDefaultsQoS(),
       callback_group_);
+    rcl_service_introspection_state_t introspection_state = RCL_SERVICE_INTROSPECTION_OFF;
+    if(!node_->has_parameter("service_introspection_mode")) {
+      node_->declare_parameter("service_introspection_mode", "disabled");
+    }
+    std::string service_introspection_mode =
+      node_->get_parameter("service_introspection_mode").as_string();
+    if (service_introspection_mode == "metadata") {
+      introspection_state = RCL_SERVICE_INTROSPECTION_METADATA;
+    } else if (service_introspection_mode == "contents") {
+      introspection_state = RCL_SERVICE_INTROSPECTION_CONTENTS;
+    }
+
+    this->client_->configure_introspection(
+        node_->get_clock(), rclcpp::SystemDefaultsQoS(), introspection_state);
+    if(spin_thread) {
+      callback_group_executor_ = std::make_shared<rclcpp::executors::SingleThreadedExecutor>();
+      callback_group_executor_->add_callback_group(callback_group_,
+          node_->get_node_base_interface());
+      executor_thread_ = std::make_unique<nav2_util::NodeThread>(callback_group_executor_);
```

Comment:
Oh this can't be here, otherwise `spin_until_future_complete` will throw exceptions each time its called since the executor is already spinning.